### PR TITLE
MIP-7: Add a print delegate

### PR DESCRIPTION
**Pull request by blockiosaurus**
This PR was made to address [MIP-7](https://github.com/metaplex-foundation/mip/blob/0983ac4e3ee6b517755ca6bc5db21784f42e76d7/mip-7.md), adding escrowless edition printing through a delegate.

This PR consists of two parts, adding a holder delegate and adding the additional delegate account to print.

* Holder Delegate - The holder delegate is a new, third type of delegate added for NFTs and pNFTs. It's allows for the holder to delegate permissions that specific to them. Currently the only Delegate Role is for Prints but it could easily be extended to other uses.

* Print Delegate Role - This delegate role allows for a delegated authority to print an edition, rather than just the holder. Multiple delegates can be set so an edition can be escrowlessly listed on several marketplaces at one. Because the unified Instruction architecture versions on args structs rather than the accounts context, I hacked the delegate in as a remaining account instead of created a new Instruction with duplicate handling. It switches based on the accounts array length to pass in an Option for the holder delegate role.

### CODE REVIEW

File: clients/rust/src/generated/accounts/holder_delegate_record.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+//! This code was AUTOGENERATED using the kinobi library.
+//! Please DO NOT EDIT THIS FILE, instead use visitors
+//! to add features, then rerun kinobi to update it.
+//!
+//! [https://github.com/metaplex-foundation/kinobi]
+//!
+
+use crate::generated::types::Key;
+use crate::hooked::HolderDelegateRoleSeed;
```
[the remark this answers is not in the document]
Comment (blockiosaurus) [reply clients/rust/src/generated/accounts/holder_delegate_record.rs:9]: Fixed!

File: programs/token-metadata/program/src/instruction/mod.rs [hunk 2]
```diff
@@ -806,24 +806,26 @@ pub enum MetadataInstruction {
...
     /// Given a token account containing the master edition token to prove authority, and a brand new non-metadata-ed mint with one token
     /// make a new Metadata + Edition that is a child of the master edition denoted by this authority token.
+    /// Account index 18, holder_delegate_record, is optional and only used for escrowless edition printing. It is fetched from the remaining accounts.
     #[account(0, writable, name="edition_metadata", desc="New Metadata key (pda of ['metadata', program id, mint id])")]
     #[account(1, writable, name="edition", desc="New Edition (pda of ['metadata', program id, mint id, 'edition'])")]
     #[account(2, writable, name="edition_mint", desc="Mint of new token - THIS WILL TRANSFER AUTHORITY AWAY FROM THIS KEY")]
     #[account(3, name="edition_token_account_owner", desc="Owner of the token account of new token")]
     #[account(4, writable, name="edition_token_account", desc="Token account of new token")]
-    #[account(5, signer, name="edition_mint_authority", desc="Mint authority of new mint")]
+    #[account(5, optional_signer, name="edition_mint_authority", desc="Mint authority of new mint")]
```
[the remark this answers is not in the document]
Comment (blockiosaurus) [reply programs/token-metadata/program/src/instruction/mod.rs:815]: You're right. I had a different problem and mistakenly made this change which was the improper fix.

File: programs/token-metadata/program/src/processor/delegate/delegate.rs [hunk 3]
```diff
@@ -48,6 +49,9 @@ impl Display for DelegateScenario {
                     "ProgrammableConfigItem".to_string()
                 }
             },
+            Self::Holder(role) => match role {
+                HolderDelegateRole::PrintDelegate => "PrintDelegate".to_string(),
+            },
```
Comment (danenbm) [inline programs/token-metadata/program/src/processor/delegate/delegate.rs:54]: Out of scope for this PR, but I think these display trait implementations are strange in that here it implements one set of strings in Pascal case, but elsewhere the roles have their own display traits with snake case.  I think some of them were used for a default RuleSet we made that used delegate scenarios, so maybe no changes needed ever, but just wanted to call out that it seems kind of inconsistent.

File: clients/js/test/printV2.test.ts [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+    owner: originalOwner.publicKey,
+    delegate: delegate.publicKey,
+  });
+
+  const digitalAssetWithToken = await fetchDigitalAssetWithAssociatedToken(
+    umi,
+    originalMint.publicKey,
+    originalOwner.publicKey
+  );
+  // const tokenAccount = await fetchToken(umi, digitalAssetWithToken.token.publicKey);
```
Comment (danenbm) [inline clients/js/test/printV2.test.ts:518]: Nit: cleanup

File: clients/js/test/printV2.test.ts [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+    editionMint.publicKey,
+    editionOwner.publicKey
+  );
+  t.like(editionAsset, <DigitalAssetWithToken>{
+    publicKey: editionMint.publicKey,
+    metadata: {
+      name: 'My NFT',
+      symbol: 'MNFT',
+      uri: 'https://example.com/nft.json',
+      sellerFeeBasisPoints: 542,
```
Comment (danenbm) [inline clients/js/test/printV2.test.ts:71]: Nit: do you also want to check the token standard here.

File: clients/js/test/printV2.test.ts [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+    owner: originalOwner.publicKey,
+    delegate: delegate.publicKey,
+  });
+
+  const digitalAssetWithToken = await fetchDigitalAssetWithAssociatedToken(
+    umi,
+    originalMint.publicKey,
+    originalOwner.publicKey
+  );
+  // const tokenAccount = await fetchToken(umi, digitalAssetWithToken.token.publicKey);
```
Comment (danenbm) [inline clients/js/test/printV2.test.ts:287]: Nit: cleanup

File: clients/js/test/printV2.test.ts [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+    },
+    edition: {
+      isOriginal: false,
+      parent: findMasterEditionPda(umi, { mint: originalMint.publicKey })[0],
+      edition: 2n,
+    },
+  });
+});
+
+test('it can print a new owner as the master holder after delegating the authority', async (t) => {
```
Comment (danenbm) [inline clients/js/test/printV2.test.ts:490]: Nit: This title could be a bit clearer.

File: programs/token-metadata/program/src/instruction/mod.rs [hunk 8]
```diff
@@ -806,24 +806,26 @@ pub enum MetadataInstruction {
...
     #[account(12, name="master_metadata", desc="Master record metadata account")]
     #[account(13, name="update_authority", desc="The update authority of the master edition.")]
     #[account(14, name="spl_token_program", desc="Token program")]
     #[account(15, name="spl_ata_program", desc="SPL Associated Token Account program")]
     #[account(16, name="sysvar_instructions", desc="Instructions sysvar account")]
     #[account(17, name="system_program", desc="System program")]
+    // #[account(18, optional, name="holder_delegate_record", desc="The Delegate Record authorizing escrowless edition printing")]
     #[args(initialize_mint: bool)]
     Print(PrintArgs),
 }
```
[the remark this answers is not in the document]
Comment (danenbm) [reply programs/token-metadata/program/src/instruction/mod.rs]: Discussed offline, going with PrintV2 via the clients.  Perhaps in the future we will update shank to be able to version accounts, but at this time not adding a PrintV2 to the program.

File: programs/token-metadata/program/src/processor/metadata/print.rs [hunk 9]
```diff
@@ -26,16 +26,49 @@ pub fn print<'a>(
...
+    let context = Print::to_context(&accounts[0..18])?;
+
+    let holder_delegate_record_info = if accounts[18].key == &crate::ID {
+        None
+    } else {
+        Some(&accounts[18])
+    };
+
+    print_logic(program_id, context, args, holder_delegate_record_info)
+}
```
[the remark this answers is not in the document]
Comment (blockiosaurus) [reply programs/token-metadata/program/src/processor/metadata/print.rs:59]: Yup!

File: clients/js/src/generated/instructions/printV2.ts [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+  }
+  if (!resolvedAccounts.editionMintAuthority.value) {
+    if (resolvedAccounts.holderDelegateRecord.value) {
+      resolvedAccounts.editionMintAuthority.value = expectSome(
+        resolvedAccounts.payer.value
+      );
+    } else {
+      resolvedAccounts.editionMintAuthority.value = context.identity;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (danenbm) [reply clients/js/src/generated/instructions/printV2.ts:268]: Discussed offline, clear now

File: programs/token-metadata/program/src/processor/delegate/delegate.rs (hunk 3 above)
Comment (danenbm) [reply programs/token-metadata/program/src/processor/delegate/delegate.rs:54]: Discussed offline

**Review by danenbm — APPROVED** [review]

**Pull request merged**